### DISCUSSION TITLE
Custom Notebook Controller and Code Cells

### DISCUSSION THREAD

**Discussion by gbrueckl** [microsoft/vscode-discussions#331] [Extension Development QnA] [answered]
I created a custom Notebook Kernel and everything works as expected. it is registered with KernelType = "jupyter-notebook" I can execute notebooks cell-by-cell etc.
The very same kernel is also registered as interactive kernel (KernelType = "interactive") to be used in the interactive window. This also works so far and I can execute commands against my custom kernel in the interactive window directly.
However, I would like to use this interactive kernel also when running Code Cells. But this only works partially. When I try to run a Code Cell from a e.g. .py file, the code from the code cell is copied to the interactive window but when VSCode tries to run the cell automatically, I get the following error:
```notebook controller is NOT associated to notebook: vscode-interactive:Interactive-1.interactive. View Jupyter [log](command:jupyter.viewOutput) for further details.```
selecting the very same cell again in the interactive window and executing it manually works again
any ideas how to get this working?

**Comment by r3m0t** [comment, marked as answer]
The code cell is a code lens from the `vscode-jupyter` extension, so when you click it code from that extension is running and it's expecting to find a notebook or interactive that it is already controlling. You've set `KernelType = "jupyter-notebook"` which is incorrect because that type is already used by the `vscode-jupyter` extension leading to conflicts.

You can either:

1. Create a Jupyter kernel that works with Jupyter. Try it using the Jupyter Notebook software (web based), and then it will automatically be compatible with VS Code as long as the `vscode-jupyter` extension is installed and the user connects to their Jupyter server. You wouldn't use the VS Code API at all. https://docs.jupyter.org/en/latest/projects/kernels.html + https://code.visualstudio.com/docs/datascience/jupyter-notebooks#_connect-to-a-remote-jupyter-server
2. Uninstall the `vscode-jupyter` extension and use the VS Code Notebook Controller API. Register a new kernelType as you wouldn't be using Jupyter at all. Also, use the VS Code Code Lens API to parse files for Code Cells and create commands like Run Cell, Debug Cell etc. It would be similar to the open source `vscode-jupyter` extension. (Your users can have `vscode-jupyter` installed, but you should uninstall it so you know what functionality you need to implement)

Option 1 is probably better as your notebook will be compatible with the Jupyter ecosystem instead of just working in VS Code. It has features like export to HTML (using nbconvert), web based interface, etc.

**Reply by gbrueckl** [reply]
thanks for the answer @r3m0t !
this does not solve my problem but actually makes a lot of sense. 
**regarding 1:** 
My notebook kernel does not do anything with jupyter at all except leveraging the same serializer. When a cell is executed, the code is sent to an API which then returns the results and displays them. So it can never work in the jupyter eco system and also I want to work in VSCode Web which I think is also not possible when using a "real" jupyter kernel as this needs to be hosted somewhere.

**regarding 2:** 
creating my own Code Lens is probably to much work as of now but I will definitely consider it fo probably to r the future!

thanks,
Gerhard
